### ISSUE TITLE
Strange parsing issue concerning G1 collector

### ISSUE THREAD

**Issue by YouriAndropov** [chewiebug/GCViewer#128]
Hi,

First of all, it would be great if gcviewer could ignore the "G1Ergonomics" messages from the -XX:+PrintAdaptiveSizePolicy setting.

Second, I do have a strange bug while parsing gc log files.

This one works fine :

2015-02-13T17:12:01.182+0100: 195,970: Total time for which application threads were stopped: 0,0001000 seconds
2015-02-13T17:12:01.265+0100: 196,053: [GC pause (G1 Evacuation Pause) (young)
Desired survivor size 20447232 bytes, new threshold 15 (max 15)
- age   1:     503304 bytes,     503304 total
- age   2:      43928 bytes,     547232 total
, 0,0331820 secs]
   [Parallel Time: 22,5 ms, GC Workers: 4]
      [GC Worker Start (ms): Min: 196053,3, Avg: 196053,3, Max: 196053,4, Diff: 0,1]
      [Ext Root Scanning (ms): Min: 11,0, Avg: 11,5, Max: 12,1, Diff: 1,1, Sum: 46,1]
      [Update RS (ms): Min: 0,6, Avg: 1,1, Max: 1,5, Diff: 0,9, Sum: 4,3]
         [Processed Buffers: Min: 12, Avg: 19,0, Max: 30, Diff: 18, Sum: 76]
      [Scan RS (ms): Min: 0,9, Avg: 1,0, Max: 1,0, Diff: 0,1, Sum: 3,8]
      [Code Root Scanning (ms): Min: 0,0, Avg: 0,0, Max: 0,1, Diff: 0,0, Sum: 0,1]
      [Object Copy (ms): Min: 8,4, Avg: 8,7, Max: 8,9, Diff: 0,5, Sum: 34,9]
      [Termination (ms): Min: 0,0, Avg: 0,0, Max: 0,0, Diff: 0,0, Sum: 0,0]
      [GC Worker Other (ms): Min: 0,0, Avg: 0,0, Max: 0,0, Diff: 0,0, Sum: 0,1]
      [GC Worker Total (ms): Min: 22,3, Avg: 22,4, Max: 22,4, Diff: 0,1, Sum: 89,4]
      [GC Worker End (ms): Min: 196075,7, Avg: 196075,7, Max: 196075,7, Diff: 0,0]
   [Code Root Fixup: 0,0 ms]
   [Code Root Migration: 0,1 ms]
   [Clear CT: 0,5 ms]
   [Other: 10,1 ms]
      [Choose CSet: 0,0 ms]
      [Ref Proc: 8,6 ms]
      [Ref Enq: 0,2 ms]
      [Free CSet: 1,0 ms]
   [Eden: 304,0M(304,0M)->0,0B(279,0M) Survivors: 3072,0K->28,0M Heap: 418,0M(512,0M)->139,3M(512,0M)]
 [Times: user=0,09 sys=0,00, real=0,03 secs] 
2015-02-13T17:12:01.299+0100: 196,087: Total time for which application threads were stopped: 0,0338710 seconds

But, this one doesn't work :

2015-02-13T17:12:01.196+0100: 195,854: Total time for which application threads were stopped: 0,0000890 seconds
2015-02-13T17:12:01.729+0100: 196,388: [GC pause (G1 Evacuation Pause) (young)
Desired survivor size 20447232 bytes, new threshold 15 (max 15)
- age   1:    6700576 bytes,    6700576 total
- age   2:   10504568 bytes,   17205144 total
- age   3:     320040 bytes,   17525184 total
- age   4:      92336 bytes,   17617520 total
, 0,0608680 secs]
   [Parallel Time: 48,4 ms, GC Workers: 4]
      [GC Worker Start (ms): Min: 196388,7, Avg: 196390,4, Max: 196394,8, Diff: 6,1]
      [Ext Root Scanning (ms): Min: 14,1, Avg: 16,5, Max: 20,1, Diff: 6,0, Sum: 66,0]
      [Update RS (ms): Min: 0,0, Avg: 0,6, Max: 2,2, Diff: 2,2, Sum: 2,2]
         [Processed Buffers: Min: 0, Avg: 20,8, Max: 83, Diff: 83, Sum: 83]
      [Scan RS (ms): Min: 0,3, Avg: 1,4, Max: 2,5, Diff: 2,2, Sum: 5,7]
      [Code Root Scanning (ms): Min: 0,0, Avg: 0,1, Max: 0,1, Diff: 0,1, Sum: 0,2]
      [Object Copy (ms): Min: 26,6, Avg: 27,4, Max: 28,5, Diff: 1,9, Sum: 109,6]
      [Termination (ms): Min: 0,0, Avg: 0,0, Max: 0,0, Diff: 0,0, Sum: 0,1]
      [GC Worker Other (ms): Min: 0,0, Avg: 0,0, Max: 0,1, Diff: 0,0, Sum: 0,2]
      [GC Worker Total (ms): Min: 41,6, Avg: 46,0, Max: 47,8, Diff: 6,2, Sum: 183,9]
      [GC Worker End (ms): Min: 196436,4, Avg: 196436,4, Max: 196436,4, Diff: 0,0]
   [Code Root Fixup: 0,1 ms]
   [Code Root Migration: 0,2 ms]
   [Clear CT: 0,3 ms]
   [Other: 11,8 ms]
      [Choose CSet: 0,0 ms]
      [Ref Proc: 10,5 ms]
      [Ref Enq: 0,3 ms]
      [Free CSet: 0,7 ms]
   [Eden: 281,0M(281,0M)->0,0B(180,0M) Survivors: 26,0M->39,0M Heap: 398,8M(512,0M)->138,5M(512,0M)]
 [Times: user=0,20 sys=0,00, real=0,06 secs] 
2015-02-13T17:12:01.791+0100: 196,450: Total time for which application threads were stopped: 0,0624330 seconds

I couldn't see any structural difference between these two samples.
Here are the error messages:

AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.ParseException: Error parsing entry (java.lang.ArrayIndexOutOfBoundsException) Line 245: 2015-02-13T17:12:01.729+0100: 196,388: [GC pause (G1 Evacuation Pause) (young)
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: '' Line 251: , 0,0608680 secs]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Parallel Time' Line 252:    [Parallel Time: 48,4 ms, GC Workers: 4]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'GC Worker Start (ms)' Line 253:       [GC Worker Start (ms): Min: 196388,7, Avg: 196390,4, Max: 196394,8, Diff: 6,1]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Ext Root Scanning (ms)' Line 254:       [Ext Root Scanning (ms): Min: 14,1, Avg: 16,5, Max: 20,1, Diff: 6,0, Sum: 66,0]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Update RS (ms)' Line 255:       [Update RS (ms): Min: 0,0, Avg: 0,6, Max: 2,2, Diff: 2,2, Sum: 2,2]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Processed Buffers' Line 256:          [Processed Buffers: Min: 0, Avg: 20,8, Max: 83, Diff: 83, Sum: 83]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Scan RS (ms)' Line 257:       [Scan RS (ms): Min: 0,3, Avg: 1,4, Max: 2,5, Diff: 2,2, Sum: 5,7]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Code Root Scanning (ms)' Line 258:       [Code Root Scanning (ms): Min: 0,0, Avg: 0,1, Max: 0,1, Diff: 0,1, Sum: 0,2]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Object Copy (ms)' Line 259:       [Object Copy (ms): Min: 26,6, Avg: 27,4, Max: 28,5, Diff: 1,9, Sum: 109,6]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Termination (ms)' Line 260:       [Termination (ms): Min: 0,0, Avg: 0,0, Max: 0,0, Diff: 0,0, Sum: 0,1]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'GC Worker Other (ms)' Line 261:       [GC Worker Other (ms): Min: 0,0, Avg: 0,0, Max: 0,1, Diff: 0,0, Sum: 0,2]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'GC Worker Total (ms)' Line 262:       [GC Worker Total (ms): Min: 41,6, Avg: 46,0, Max: 47,8, Diff: 6,2, Sum: 183,9]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'GC Worker End (ms)' Line 263:       [GC Worker End (ms): Min: 196436,4, Avg: 196436,4, Max: 196436,4, Diff: 0,0]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Code Root Fixup' Line 264:    [Code Root Fixup: 0,1 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Code Root Migration' Line 265:    [Code Root Migration: 0,2 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Clear CT' Line 266:    [Clear CT: 0,3 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Other' Line 267:    [Other: 11,8 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Choose CSet' Line 268:       [Choose CSet: 0,0 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Ref Proc' Line 269:       [Ref Proc: 10,5 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Ref Enq' Line 270:       [Ref Enq: 0,3 ms]
AVERTISSEMENT [DataReaderSun1_6_0]: com.tagtraum.perf.gcviewer.imp.UnknownGcTypeException: Unknown gc type: 'Free CSet' Line 271:       [Free CSet: 0,7 ms]

Any clue ?

**Comment by chewiebug** [comment]
What version of gcviewer are you using? G1ergonomics should be ignored by the latest version. As far as the other problem is concerned, I'll have to look into the details. 


-------- Original Message --------

**Comment by YouriAndropov** [comment]
The latest version has this issue (1.34.1) but also the previous one (1.34)

**Comment by chewiebug** [comment]
Hi Youri,

I can't reproduce the issue with the samples you posted. Both parse 
without any warnings. Could you please send the (compressed) original 
log file to gcviewer-info@googlegroups.com?

Best regards,
Jörg

**Comment by YouriAndropov** [comment]
I had to use hotspot JVM 1.8 with GCviewer. Could it be related to this issue?
I'll send the logs tomorrow.

**Comment by YouriAndropov** [comment]
I've sent you a log not working (not the one I told about).
I think there are some regressions with latest versions of GCviewer, 1.34.1 cannot work with java 1.7 and has parsing issues on more files than the previous version.

**Comment by chewiebug** [comment]
I've received a log file via gcviewer-info mailinglist with a sender 
named "Frédéric Girard". Is that the one you sent?

Concerning GCViewer and java 1.7 please read the release notes 
<https://github.com/chewiebug/GCViewer/wiki/Changelog#gcviewer-1341---31012015-changes-since-134> 
(there was a switch to java 8).

If you have examples for those parsing issues you mention, please send 
them. I have a unittest suite, that always is run before releases. It 
may be, that some tests are missing.

**Comment by chewiebug** [comment]
Hi Frederic,

The problem is, that GCViewer fails to detect for this particular file, 
that it is a G1 gc log. It tries to parse it as a CMS log, that's where 
alle the warnings come from.

As a workaround, you can delete the first (few) lines of your log. Then 
recognition and parsing will work.

I am working on a fix to recognise the original file correctly.

Best regards,
Jörg

**Comment by YouriAndropov** [comment]
Yes that was me.
The gc log is already "cut" to a particular event, I cannot remove any lines without affecting values like total pause, throughput etc.
Surprisingly, the detection works often, but not always. Could I add a header to help the detection work ?

**Comment by chewiebug** [comment]
For recognition of the log type, GCViewer reads the first 4 KB of a file and tries to find a pattern, allowing it to decide on the type (for the details see https://github.com/chewiebug/GCViewer/blob/develop/src/main/java/com/tagtraum/perf/gcviewer/imp/DataReaderFactory.java#L92). 

There was a bug, that sometimes resulted in a wrong decision. I have fixed it for your current sample file. Recognition should generally improve now, also for manually cut files.

If it still doesn't recognise correctly, you may add " (young)" (without quotes) as the first line in a G1 log file. Then it should always be recognised as such.

I'd be interested in the first few lines of other files, that can't be recognised, however - maybe I can further improve recognition.

Actually, you are waiting for #5, right?

**Comment by chewiebug** [comment]
I have uploaded the improved version (see [Changelog](https://github.com/chewiebug/GCViewer/wiki/Changelog)). Does it work better for your files?

**Comment by YouriAndropov** [comment]
I'll try it as soon as sourceforge downloads are back ...
The feature request #5 seems very useful indeed !

**Comment by YouriAndropov** [comment]
Hello, I still have the issue with some logs, and adding a line with "(young)" doesn't work.
I've sent it by mail.
Maybe you should do a 2-pass parsing, not just parsing 4 KB of log to determine the gc type.

**Comment by chewiebug** [comment]
For the workaround to work, you need to include a space before " (young)". 

For your latest log, recognition is OK. Now the problem is with G1ergonomics. The format seems changed. I'll have to look into this. 

What JDK version are you using?

**Comment by YouriAndropov** [comment]
$ java -version
java version "1.7.0_67"
Java(TM) SE Runtime Environment (build 1.7.0_67-b01)
Java HotSpot(TM) 64-Bit Server VM (build 24.65-b04, mixed mode)

**Comment by chewiebug** [comment]
Thanks!

The problem was with the localised format of the initial timestamp in 
lines like the following:

` 16578,506: [G1Ergonomics (CSet Construction) start choosing CSet, 
_pending_cards: 1887, predicted base time: 16,71 ms, remaining time: 
33,29 ms, target pause time: 50,00 ms]`

The "," in the timestamp was not expected. It now is and the 2nd log you 
sent, is parsed fine now (please download latest 1.35-SNAPSHOT 
<http://sourceforge.net/projects/gcviewer/files/gcviewer-1.35-SNAPSHOT.jar/download> 
version).

Are there other logfiles, where recognition still does not work?

**Comment by YouriAndropov** [comment]
Not yet, but I'll tell you if it happens.
By the way, I do have some permgen OOM issues, and I just noticed that GCviewer don't graph permgen space usage. Do you plan to add some permgen metrics analysis ?

**Comment by YouriAndropov** [comment]
Hi,
I do have a bug: a negative min pause time. I've sent the gc log.

**Comment by chewiebug** [comment]
Graphing permgen space usage would be #15.

I have pushed a fix for your negative pause. It is available as the latest [1.35-SNAPSHOT](http://sourceforge.net/projects/gcviewer/files/gcviewer-1.35-SNAPSHOT.jar/download) version (2015-03-01T14:43:09+0100).

For documentation reasons here the details about this bug:
The time written in the log for "application stopped" events always includes the GC pause time of the event immediately before. If there is none, then the "application stopped" time includes only non GC related time. GCViewer only records the "overhead" of an "application stopped" event (time spent outside the GCEvent).

For this bug, the "application stopped" event immediately after the last GC pause could not be parsed. The next "application stopped" event, that could be parsed, was shorter than the previous GC pause time, hence the negative time.

**Comment by YouriAndropov** [comment]
great, thank you!

**Comment by YouriAndropov** [comment]
Another great feature could be to generate a PNG of the current graphs. From command line, we could choose the picture size (height and width).

**Comment by YouriAndropov** [comment]
Ok I do have another parsing error, with a MarkSweepCompact GC. See your mails.

**Comment by chewiebug** [comment]
I have fixed the issue. The combination of the flags 
-XX:+PrintAdaptiveSizePolicy, -XX:+PrintTenuringDistribution, 
-XX:+PrintApplicationStoppedTime used with parallel scavenge algorithm 
could not be handled by the parser. Actually, the statistics were 
correct also with the error messages, but that's probably more luck than 
planning ;-).

Regarding the PNGs for current graphs: I assume, you know, that from 
command line mode, a PNG can be created for a file:

java -jar gcviewer-1.3x.jar gc.log summary.csv [chart.png] [-t PLAIN|CSV|CSV_TS|SIMPLE|SUMMARY]

The size of the graph is taken from %user-home%/gcviewer.properties 
(window.height + window.width).

I think, with your request to create a PNG from the current graph, and 
be able to set the graph size in command line mode, you finally hit a 
spot with no open issue yet :-). Will you open one?

**Comment by YouriAndropov** [comment]
Does the last snapshot download include your fix ?

**Comment by chewiebug** [comment]
It should do so, yes.

**Comment by fedragon** [comment]
Hello @chewiebug, 
thank you for keeping this tool up to date, it is a great help!

I'm feeding gcviewer a gc.log file created by G1 GC and I get some parsing errors which resemble the ones described in this issue. 
I just cloned this repo and built from source so I assume your fix to be already in. 
Can I send the detected errors and the log file to gcviewer-info@googlegroups.com?

**Comment by chewiebug** [comment]
Yes, please send them to the mailing list. 


-------- Original Message --------

**Comment by fedragon** [comment]
thank you, I just sent a mail with the log at fault.

**Comment by chewiebug** [comment]
Hi Federico,

Unfortunately, this is also one of those, that is so hard to parse, that 
currently I have no solution to it.

Here is, what a normal GC event looks like (<timestamp, name, duration> 
all in one line):
2015-05-01T11:26:29.623+0200: 160770.809: [GC pause (G1 Humongous 
Allocation) (young) (initial-mark), 0.0625887 secs]
    [Parallel Time: 60.5 ms, GC Workers: 10]
    ...

Here is, what the parser encounters (<timestamp, name, [two concurrent 
events], duration> - spread over 3 lines):
2015-05-01T11:26:29.687+0200: 160770.873: [GC pause (G1 Humongous 
Allocation) (young)2015-05-01T11:26:29.701+0200: 160770.887: [GC 
concurrent-root-region-scan-end, 0.0157002 secs]
2015-05-01T11:26:29.701+0200: 160770.887: [GC concurrent-mark-start]
, 0.0470136 secs]
    [Root Region Scan Waiting: 14.6 ms]
    [Parallel Time: 23.4 ms, GC Workers: 10]
    ...

The parser usually can cope with one concurrent collection in the middle 
of a GC pause, but two are too many - and very rare to occur (only once 
in your log with about 750'000 lines).

Sorry to disappoint you!

Best regards,
Jörg

**Comment by chewiebug** [comment]
Oops, mailing to the github issue killed the coloring and the 
linebreaks. It looks better in the mailing list: 
https://groups.google.com/forum/#!topic/gcviewer-info/bL4hteqH_eo

**Comment by fedragon** [comment]
Hi Jörg,
I understand, thank you for looking into it anyway!

Just one more question: will gcviewer try to parse the whole file or stop at the first error? I see the graph going flat but I was wondering if I could still rely on the statistics.

**Comment by chewiebug** [comment]
Hi Federico,

You are welcome!

To your question: GCViewer tries to parse every line of a gc log. If an 
error is encountered, it continues with the next line. Sometimes a few 
lines are skipped, if they are known to contain no information, that 
GCViewer evaluates - but that mechanism is very defensive.

In the case of the log you sent, GCViewer counts 26 parsing problems, 
but only two events are missed ("GC pause (G1 Humongous Allocation) 
(young)" and "GC concurrent-mark-start". "GC 
concurrent-root-region-scan-end" is found and added to the total number 
of events).

Most of the 26 parsing problems are actually part of the "GC pause (G1 
Humongous Allocation) (young)" event, which contains lots of detail 
information about the steps that were done. Since the beginning of the 
event couldn't be recognised, the parser tries to find usable 
information in all those detail information, that it doesn't recognise, 
either. The parser recovers with the first line after that event and 
parses it fine again.

Best regards,
Jörg

**Comment by fedragon** [comment]
Cool, thank you!

**Comment by chewiebug** [comment]
I have pushed a fix, that will reduce the number of warning messages, when one of these situations are encountered.

Otherwise, I think, I can't do much more to improve the parser. So I close this issue.
